File: examples/canvas_shapes.nim
## Canvas Shapes Example
##
## Demonstrates the Canvas API with various drawing operations.
## Shows how to draw rectangles, circles, lines, and use different colors.

import std/math
import ../src/kryon

# Drawing procedure that creates various shapes using high-level drawing functions
proc drawShapes(ctx: DrawingContext, width, height: float) =
  # Clear canvas with light blue background
  ctx.fillStyle = rgba(135.uint8, 206.uint8, 235.uint8, 255.uint8)  # Sky blue
  ctx.drawRectangle(0, 0, width, height)

  # Draw a red filled rectangle
  ctx.fillStyle = rgba(255.uint8, 0.uint8, 0.uint8, 255.uint8)  # Red
  ctx.drawRectangle(50, 50, 100, 80)

  # Draw a blue filled circle
  ctx.fillStyle = rgba(0.uint8, 0.uint8, 255.uint8, 255.uint8)  # Blue
  ctx.drawCircle(250, 90, 40)

  # Draw a green filled triangle
  ctx.fillStyle = rgba(0.uint8, 255.uint8, 0.uint8, 255.uint8)  # Green
  ctx.drawTriangle(400, 130, 350, 50, 450, 50)

  # Draw a purple rectangle outline
  ctx.strokeStyle = rgba(128.uint8, 0.uint8, 128.uint8, 255.uint8)  # Purple
  ctx.lineWidth = 3.0
  ctx.drawRectangleLines(50, 200, 150, 100)

  # Draw an orange circle outline
  ctx.strokeStyle = rgba(255.uint8, 165.uint8, 0.uint8, 255.uint8)  # Orange
  ctx.lineWidth = 2.0
  ctx.drawCircleLines(300, 250, 50)

  # Draw a gold star
  ctx.fillStyle = rgba(255.uint8, 215.uint8, 0.uint8, 255.uint8)  # Gold
  ctx.drawStar(450, 250, 40.0, 20.0, 5)

  # Draw some decorative shapes
  # Pentagon
  ctx.fillStyle = rgba(255.uint8, 20.uint8, 147.uint8, 255.uint8)  # Deep Pink
  ctx.drawPoly((x: 100.0, y: 320.0), 5, 25.0)

  # Hexagon outline
  ctx.strokeStyle = rgba(0.uint8, 128.uint8, 0.uint8, 255.uint8)  # Green
  ctx.lineWidth = 2.0
  ctx.drawPolyLines((x: 200.0, y: 320.0), 6, 25.0)


  # Draw a small ellipse
  ctx.fillStyle = rgba(255.uint8, 140.uint8, 0.uint8, 255.uint8)  # Dark Orange
  ctx.drawEllipse(400, 340, 30, 15)

  # Draw a ring
  ctx.fillStyle = rgba(0.uint8, 100.uint8, 200.uint8, 255.uint8)  # Navy
  ctx.drawRing(500, 340, 10, 20)

  # Draw a sine wave
  ctx.strokeStyle = rgba(148.uint8, 0.uint8, 211.uint8, 255.uint8)  # Medium Violet Red
  ctx.lineWidth = 2.0
  ctx.drawSineWave(50, 380, 500, 15, 0.08)

  # Draw small filled polygons
  ctx.fillStyle = rgba(50.uint8, 205.uint8, 50.uint8, 255.uint8)  # Lime Green
  ctx.drawPoly((x: 150.0, y: 370.0), 4, 12.0)  # Square as polygon
  ctx.drawPoly((x: 250.0, y: 370.0), 3, 12.0)  # Triangle as polygon


# Define the UI with multiple canvases
let app = kryonApp:
  Header:
    windowWidth = 800
    windowHeight = 600
    windowTitle = "Canvas Shapes Example"

  Container:
    width = 800
    height = 600
    backgroundColor = "#2C3E50"

    # Title
    Text:
      text = "Canvas Drawing Examples"
      fontSize = 32
      color = "#ECF0F1"
      posX = 200
      posY = 20

    # First canvas - basic shapes
    Center:
      Canvas:
        width = 550
        height = 400
        backgroundColor = "#34495E"
        onDraw = drawShapes

File: examples/lua/canvas_shapes.lua
-- Canvas Shapes Example
-- Demonstrates the Love2D-style Canvas API with various drawing operations

local kryon = require("../../../bindings/lua/init")
local canvas = require("../../../bindings/lua/canvas")

-- Drawing procedure that creates various shapes using Love2D-style API
local function drawShapes()
    -- Clear canvas with light blue background
    canvas.background(135, 206, 235)  -- Sky blue

    -- Row 1: Basic filled shapes
    canvas.fill(255, 0, 0, 255)  -- Red
    canvas.rectangle(canvas.FILL, 50, 50, 80, 60)

    canvas.fill(0, 0, 255, 255)  -- Blue
    canvas.circle(canvas.FILL, 200, 80, 35)

    canvas.fill(0, 255, 0, 255)  -- Green
    -- Triangle as polygon
    local triangle_vertices = {
        {x = 300, y = 120},
        {x = 260, y = 40},
        {x = 340, y = 40}
    }
    canvas.polygon(canvas.FILL, triangle_vertices)

    -- Row 2: Arc demos (half circle and quarter circle)
    canvas.fill(255, 255, 0, 255)  -- Yellow
    canvas.arc(canvas.FILL, 450, 80, 35, 0, math.pi)  -- Half circle

    canvas.stroke(0, 255, 255, 255)  -- Cyan
    canvas.strokeWeight(3)
    canvas.arc(canvas.LINE, 80, 200, 35, 0, math.pi / 2)  -- Quarter circle

    -- Row 3: Sine wave
    canvas.stroke(148, 0, 211, 255)  -- Purple
    canvas.strokeWeight(2)
    local wave_points = {}
    for i = 0, 100 do
        local x = 50 + i * 5
        local y = 320 + math.sin(0.08 * i) * 20
        table.insert(wave_points, {x = x, y = y})
    end
    canvas.line(wave_points)
end

-- Define the UI with a canvas
local app = kryon.app({
    width = 800,
    height = 600,
    title = "Canvas Shapes Example",

    body = kryon.container({
        width = 800,
        height = 600,
        background = 0x2C3E50,  -- Dark blue-gray

        children = {
            -- Title
            kryon.text({
                text = "Canvas Drawing Examples",
                x = 20,
                y = 20,
                color = 0xECF0F1  -- Light gray
            }),

            -- Canvas with shapes
            kryon.container({
                width = 550,
                height = 400,
                background = 0x34495E,  -- Darker blue-gray
                x = 100,
                y = 100,

                children = {
                    kryon.canvas_enhanced({
                        width = 550,
                        height = 400,
                        background = 0x3498DB,
                        on_draw = drawShapes  -- No ctx parameter needed!
                    })
                }
            })
        }
    })
})

-- TODO: Run the application when runtime is implemented
-- app:run()

print("Canvas Shapes Lua example created successfully!")
print("Window: 800x600, Title: 'Canvas Shapes Example'")
print("Features: Canvas with various drawing operations")
print("Shapes: Rectangle, circle, triangle, arcs, sine wave")
print("Colors: Red, blue, green, yellow, cyan, purple")